src/main/kotlin/tanvd/grazi/ide/ui/GraziRulesTree.kt

```diff
@@ -0,0 +1,279 @@
+package tanvd.grazi.ide.ui
+
+import com.intellij.ide.CommonActionsManager
+import com.intellij.ide.DefaultTreeExpander
+import com.intellij.ide.ui.search.SearchUtil
+import com.intellij.openapi.actionSystem.ActionManager
+import com.intellij.openapi.actionSystem.DefaultActionGroup
+import com.intellij.packageDependencies.ui.TreeExpansionMonitor
+import com.intellij.ui.*
+import com.intellij.ui.SimpleTextAttributes.REGULAR_ATTRIBUTES
+import com.intellij.ui.SimpleTextAttributes.REGULAR_BOLD_ATTRIBUTES
+import com.intellij.util.ui.JBUI
+import com.intellij.util.ui.UIUtil
+import com.intellij.util.ui.tree.TreeUtil
+import org.languagetool.rules.Category
+import org.languagetool.rules.Rule
+import org.picocontainer.Disposable
+import tanvd.grazi.GraziConfig
+import tanvd.grazi.language.Lang
+import tanvd.grazi.language.LangTool
+import java.awt.BorderLayout
+import java.util.*
+import javax.swing.JPanel
+import javax.swing.JTree
+import javax.swing.SwingUtilities
+import javax.swing.tree.DefaultMutableTreeNode
+import javax.swing.tree.DefaultTreeModel
+import kotlin.Comparator
+import kotlin.collections.HashSet
+
+private data class RuleWithLang(val rule: Rule, val lang: Lang, val enabled: Boolean, var state: Boolean) {
+    val category: Category
+        get() = rule.category
+}
+
+private typealias RulesMap = SortedMap<Lang, SortedMap<Category, MutableList<RuleWithLang>>>
+
+private fun LangTool.allRulesWithLangs(): RulesMap {
+    val result = TreeMap<Lang, SortedMap<Category, MutableList<RuleWithLang>>>()
+    GraziConfig.state.enabledLanguages.forEach { lang ->
+        val categories = TreeMap<Category, MutableList<RuleWithLang>>(Comparator<Category> { o1, o2 -> o1!!.name.compareTo(o2!!.name) })
+        val activeRules = this[lang].allActiveRules.toHashSet()
+        val usedRules = HashSet<String>() // prevent subrules with same ids
+
+        this[lang].allRules.forEach {rule ->
+            if (rule.id !in usedRules) {
+                usedRules.add(rule.id)
```

Comment:
Use instead unneeded hashset `distinctBy { it.id }`

Reply:
Done